subprojects/core/src/test/groovy/org/gradle/initialization/DefaultExceptionAnalyserTest.groovy

```diff
@@ -0,0 +1,273 @@
+ * limitations under the License.
+ */
+package org.gradle.initialization
+
+import org.gradle.api.GradleScriptException
+import org.gradle.api.tasks.TaskExecutionException
+import org.gradle.groovy.scripts.Script
+import org.gradle.groovy.scripts.ScriptSource
+import org.gradle.internal.event.ListenerManager
+import org.gradle.internal.exceptions.Contextual
+import org.gradle.internal.exceptions.LocationAwareException
+import org.gradle.internal.exceptions.MultiCauseException
+import spock.lang.Specification
+
+class DefaultExceptionAnalyserTest extends Specification {
+
+    private final ListenerManager listenerManager = Mock(ListenerManager.class)
+    private final StackTraceElement element = new StackTraceElement("class", "method", "filename", 7)
+    private final StackTraceElement callerElement = new StackTraceElement("class", "method", "filename", 11)
+    private final StackTraceElement otherElement = new StackTraceElement("class", "method", "otherfile", 11)
+    private final StackTraceElement elementWithNoSourceFile = new StackTraceElement("class", "method", null, 11)
+    private final StackTraceElement elementWithNoLineNumber = new StackTraceElement("class", "method", "filename", -1)
+    private final ScriptSource source = Mock(ScriptSource.class)
+
+    def setup() {
+        source.getFileName() >> "filename"
+        source.getDisplayName() >> "build file filename"
+    }
+
+    def 'wraps original exception when it is not a contextual exception'() {
+        given:
+        Throwable failure = new RuntimeException()
+        DefaultExceptionAnalyser analyser = analyser()
+
+        when:
+        Throwable transformed = analyser.transform(failure)
+
+        then:
+        transformed instanceof LocationAwareException
+
+        LocationAwareException gse = (LocationAwareException) transformed
+        gse.getCause().is(failure)
+        gse.getReportableCauses().isEmpty()
+    }
+
+    def 'wraps contextual exception with location aware exception'() {
+        given:
+        Throwable failure = new ContextualException()
+        DefaultExceptionAnalyser analyser = analyser()
+
+        when:
+        Throwable transformedFailure = analyser.transform(failure)
+
+        then:
+        transformedFailure instanceof LocationAwareException
+
+        LocationAwareException gse = (LocationAwareException) transformedFailure
+        gse.getCause().is(failure)
+        gse.getReportableCauses().isEmpty()
+    }
+
+    def 'wraps highest contextual exception with location aware exception'() {
+        given:
+        Throwable cause = new ContextualException()
+        Throwable failure = new ContextualException(cause)
+        DefaultExceptionAnalyser analyser = analyser()
+
+        when:
+        Throwable transformedFailure = analyser.transform(failure)
+
+        then:
+        LocationAwareException gse = (LocationAwareException) transformedFailure
+        gse.getCause()is(failure)
```

Comment:
`gse.cause.is(failure)` would be more readable (lack of space or dot in your code)